File: 0001-two-sum/0001-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */


var twoSum = function(nums, target) {
    const numMap = new Map(); // Create a Map to store numbers and their indices

    for (let i = 0; i < nums.length; i++) {
        const currentNum = nums[i];

        const complement = target - currentNum;

        // Check if the complement already exists in our map
        if (numMap.has(complement)) {
            // If it exists, we found our two numbers
            // numMap.get(complement) gives the index of the complement
            // i is the index of the current number
            return [numMap.get(complement), i];
        }

        numMap.set(currentNum, i);
    }

    return [];
};

// var twoSum = function(nums, target) {
//    var new_nums = nums
//    for(var i=0; i<nums.length; i++) {
//     console.log(`before shift: ${nums}`)
//        var x = nums[0]
//        console.log(`x: ${x}`)
//        nums.shift()
//        console.log(`after shift: ${nums}`)        

//        for(var j=0; j<nums.length; j++){
//             var res = target-x
//             console.log(`res: ${res}`)
//             if(nums.includes(res)){
//                 return [i,i+j+1]
//             }
//        }
       
//    }
// };

File: 0001-two-sum/0001-two-sum.go
func twoSum(nums []int, target int) []int {
    numMap := make(map[int]int)

	for i, currentNum := range nums {
		// Calculate the complement needed to reach the target.
		complement := target - currentNum

		// Check if the complement already exists in our map.
		// The `ok` variable will be true if the key exists, and `idx` will be its value.
		if idx, ok := numMap[complement]; ok {
			// If the complement exists, we've found our two numbers.
			// Return the index of the complement (idx) and the current number's index (i).
			return []int{idx, i}
		}

		// If the complement is not found, add the current number and its index to the map.
		numMap[currentNum] = i
	}

	// If no solution is found after iterating through the entire slice,
	// return an empty slice as per the problem's typical requirements.
	return []int{}
}